rfcs/000-caller-side-managed-buffers.md

```diff
@@ -0,0 +1,833 @@
+
+# Motivation / Background
+
+[Issue 1362][gh1362] lists some issues with the existing `{Client,Server}Connection` API but to summarize:
+
+[gh1362]: https://github.com/rustls/rustls/issues/1362
+
+- the internal buffering results in extra memcpy-ing
+- because the buffers are internal they can't be resized nor deallocated, when they are no longer needed, by the caller / end-user
+- the API is not no-std compatible because it uses the `io::{Read,Write}` traits which are defined in libstd
+
+The `Connection` API in this proposal addresses the above issues by
+
+- requiring the end user to manage all buffers -- thus end-users can resize, allocate, deallocate the buffers as they see fit
+- not including the `io` traits in its API -- thus the API has no-std support out of the box
+- being state machine like to support both blocking and async IO
+
+Lastly, the current `Connection` API internally / implicitly performs certificate validation.
+When verification is done using `rustls_webpki`, this is fine because that API does not perform any IO action.
+However, in the future when using `rustls-platform-verifier` with `rustls` becomes possible,
+implicit verification becomes an issue because, on some platforms, `rustls-platform-verifier` performs IO actions like file operations or even network operations in some cases (e.g. on macOS).
+
+# Guide-level explanation
+
+> **NOTE**: a mock implementation, with runnable tests / examples, of the API presented in this section can be found in the `sketch` directory located in the root of this branch
+
+This section explains the new API from the point of view of an end user.
+
+## Overview
+
+The top level API are the `LlClientConnection` and `LlServerConnection` structs.
+These are lower level version of the existing `{Client,Server}Connection` API.
+Their constructors will pretty much mimic the constructors of the existing `Connection` types.
+
+```rust
+impl LlClientConnection {
+    pub fn new(config: Arc<ClientConfig>, name: ServerName) -> Result<Self, Error> {
+        // ..
+    }
+}
+
+impl LlServerConnection {
+    pub fn new(config: Arc<ServerConfig>) -> Result<Self, Error> {
+        // ..
+    }
+}
+```
+
+The differences start in the way how IO is performed.
+The `LlConnection` types do not operate on IO objects; they operate on buffers.
+The main methods of the `LlConnection` types are shown below:
+
+```rust
+type Result<T> = core::result::Result<T, rustls::Error>;
+
+// both `LlClientConnection` and `LlServerConnection` implement `DerefMut<Target = LlConnectionCommon>`
+pub struct LlConnectionCommon { /* .. */ }
+
+impl LlConnectionCommon {
+    /// Handles TLS records
+    ///
+    /// This method must always be called and its returned `Status` checked prior to
+    /// calling either `encrypt_outgoing` or `decrypt_incoming`
+    pub fn handle_tls_records<B>(
+        &mut self,
+        incoming_tls: &mut IncomingTls<B>,
+        outgoing_tls: &mut Vec<u8>,
```

Comment:
`&mut Vec` is not great, as it relies on `alloc`.  Some embedded systems might not have `alloc`, or may only have a subset of `alloc` that only includes fallable APIs.

Reply:
Yeah, I think we should be relying on `&mut [u8]` for these APIs. If `Status::state` is going to describe which actions are allowed/possible, then it feels like `Status` should hold references to the `LlConnectionCommon`, the `incoming_tls` and `outgoing_tls` and expose `encrypt_outgoing()` and `decrypt_incoming()` on a type exposed via the `State` variant.

I think this would be simpler if we only offered a method like `pub fn process_new_packets(&mut self, incoming_tls: &mut [u8], send_plaintext: &[u8], outgoing_tls: &mut [u8])` which would then yield, among other things, an iterator over `&[u8]` blocks of plaintext data pointing into `incoming_tls`.

Reply:
@djc thanks for the feedback!

> I think this would be simpler if we only offered a method like

I made a new sketch (`sketch2`) using this idea and I like the direction. I suggest you check the compile-pass test first: https://github.com/japaric/rustls/blob/d21cb4f7e2e79a64e447e0e147cce5cb13148347/sketch2/src/lib.rs#L174

I pushed the sketch into this branch / PR so you can also see it in this PR webpage.

what I ended up with is:
- `LlConnectionCommon` only has one method
- `process_new_packets` / `process_tls_records` only takes a single argument: `incoming_tls`
- `outgoing_tls` and `send_plaintext` only appear in the methods of some `State` variants. I personally like this better over storing `outgoing_tls` in `State` as that increases its lifetime parameters. I think it would also make harder to resize `outgoing_tls` while handling a variant that requires encrypting some data
- also changed `outgoing_tls` from a `Vec` into a `&mut [u8]`. this means that encrypting can now return an `Err`or due to insufficient buffer size. there's an example of resizing `outgoing_tls` buffer in the linked compile-pass test
- changed the `AppDataReceived` state to yield a single record / slice instead of an iterator. I'm still a bit undecided over which is better (slice vs iterator) but having the `process_` method handle one record on every invocation feels consistent. (I think one record per invocation fits better with the `discard` idea mentioned below)

in the signature of the `process_new_packets` method you posted, `incoming_tls` has type `&mut [u8]`. would you prefer that over providing the `IncomingTls` abstraction as per the current RFC text? that would push the task of removing TLS records from the front of the buffer to the user. I think that could be more flexible but also make the API a bit more annoying to use.

in that case, I think `process_` can return a `struct Status<_> { discard: usize, state: State<_> }` value where `discard` indicates how many bytes from the front of `incoming_tls` must be discarded before the next call to `process_`. I *think* `&mut [u8]` + `discard` would also work with libstd's `read_buf` API (`BorrowedBuf`) without needing to have those libstd types in the API which would keep the API the same in both std and no-std mode and that'd be a plus.

Reply:
> * `outgoing_tls` and `send_plaintext` only appear in the methods of some `State` variants. I personally like this better over storing `outgoing_tls` in `State` as that increases its lifetime parameters. I think it would also make harder to resize `outgoing_tls` while handling a variant that requires encrypting some data

This seems potentially problematic: what if I haven't received any TLS but I would like to send some plaintext now?

> * also changed `outgoing_tls` from a `Vec` into a `&mut [u8]`. this means that encrypting can now return an `Err`or due to insufficient buffer size. there's an example of resizing `outgoing_tls` buffer in the linked compile-pass test

Note there is some tricky stuff here: handshake states need to write their outgoing TLS **before** manipulating the connection state so we can give the caller a chance to try again with a bigger buffer, without ending up with connection state that's inconsistent with the messages processed.

> * changed the `AppDataReceived` state to yield a single record / slice instead of an iterator. I'm still a bit undecided over which is better (slice vs iterator) but having the `process_` method handle one record on every invocation feels consistent. (I think one record per invocation fits better with the `discard` idea mentioned below)

It feels like performance would be better if we can handle multiple records per invocation? Not sure though, maybe one record per invocation would be fine.

Reply:
> I think the scenario were that would happen is [early (0-RTT) data](https://datatracker.ietf.org/doc/html/rfc8446#section-2.3). the client can send early data encrypted with a shared key before receiving a record from the server. In other cases, both sides need to agree on a key before application data can be encrypted, as far as I've seen in TLS logs.

I meant just after the handshake.

Reply:
> I think this is covered. whenever a handshake record needs to be encrypted `process_tls_record` returns the `MustEncryptTlsData`. this variant has that fallible `encrypt` method; only when that method succeeds the state machine advances. if `encrypt` fails due to insufficient space, `encrypt` can be called again without issues. this is (mock) tested in `sketch2`: in one of the tests, `outgoing_tls` starts as a `Vec` with 0-capacity so the `encrypt` failure path is hit during the execution of the test.

Hmm, okay, I think that makes sense.

Reply:
> I meant just after the handshake.

once the handshake is done, `process_tls_records` returns the state variant `TrafficTransit` and that has an `encrypt` method so encryption is possible before post-handshake app-data records are received. this should be clearer once the RFC text is updated to use the `sketch2` API as the main proposal, which I'll get to shortly.